In [22]:
import pandas as pd
from mod06_deanonymize import load_data, link_records, deanonymization_rate

In [23]:
anon_path = "mod06_data/anonymized.csv"
aux_path = "mod06_data/auxiliary.csv"

### Load and review the data

In [24]:
anon, aux = load_data(anon_path, aux_path)

In [25]:
anon.head(20)

,anon_id,age,zip3,gender
0,ERM5C,50,432,F
1,OFJJP,75,323,F
2,I3H24,44,323,M
3,3WUYW,66,323,M
4,ZN91P,18,323,M
5,7BLKV,36,606,F
6,4SJFH,70,890,M
7,0NO3L,31,110,M
8,U060Z,80,191,F
9,1761H,58,606,M


In [26]:
import importlib
import mod06_deanonymize
importlib.reload(mod06_deanonymize)
from mod06_deanonymize import load_data, link_records, deanonymization_rate

### Merge the datasets by age, gender, and zip3 and return the unique matches

In [27]:
matches = link_records(anon, aux)
matches.head()

,anon_id,matched_name
0,ERM5C,Grace Lopez
7,ZN91P,Kyle Martinez
14,0NO3L,Marcus Lopez
15,U060Z,Sally Davis
16,1761H,Frank Jackson


### Calculate the match rate

In [28]:
rate = deanonymization_rate(matches, anon)
print(rate)

0.421


# Discussion Questions

### Describe your reaction to the match rate. What sources of data might be prone to privacy issues?

The match rate of 0.421 or 41% was definitely very shocking because it shows that even if you remove variables such as their name, you can still figure out who they are by using other parameters such as age, gender, and zip code. Individuals that want to identify themselves as anonymous are not truly secure as it could be very likely other corporations/individuals are collecting outside data. Health records and other location data are all really vulnerable to this kind of attack because they're often public or can be found easily.

### Provide examples of at least two other variables that could increase the match rate if included.

Two variables that could increase the match rate if included is birth date and the full zip code, instead of just the first three digits. This data would make it significantly easier for someone to match info to a identified person with other datasets. Other variables such as occupation and occupation location would also make it easier to identify an individual. 